### PR TITLE
Add a class diagram for terms.

### PR DESCRIPTION
**Pull request by aucampia**
The diagram is based on the diagram created by
Graham Higgins (@gjhiggins) in https://github.com/RDFLib/rdflib/issues/1526

This shows the class heirarchy of various terms such as Identifier,
IdentifiedNode, URIRef, Literal, etc.

The diagram is in [plantuml](https://plantuml.com/class-diagram) and
compiled to svg by the
[kroki extension for sphinx](https://github.com/sphinx-contrib/kroki).

Diagrams can be rendered from the plantuml at https://kroki.io/.

Other changes:
 - Some updates to the "Writing RDFLib Documentation" page.

### CODE REVIEW

**Comment by aucampia** [comment]
Preview:

screencapture-127-0-0-1-8080-rdf-terms-html-2022-01-22-01_00_04

**Comment by aucampia** [comment]
The diagram itself can be seen [diagram](http://www.plantuml.com/plantuml/uml/dL9DRzGm4BtxL_YuIBidRZrKvOEYIXMWHN11g9h4ayHQUnp64qee-zz9xRLgYI2jwaFal7b-7vQleW3Bv9sA1qiZC7WT1p3XtbAlEt0HguaF5Dg1WqSjFDMJ4Te3zFXPPeUQiHMWtjM4PY7q72OoRu8BhAyVrpkbMWSnwW_1ODdV62IndKNM3qel2tziWWp8B_JkKZSXk8JIEfpwn7ihmsPaxEolYmNYRSQt6iayyO-gyjfDu-xFvNeIyuYUqaZuDCiGQ9Dq5kqiKnEpL5Ma27xPt-onEmdHoms1npX2YtSU27haefa81Zk6FllaaoVuBXw2Gsz_Onw-Jg6FAfF-iJTfN3AyHx96fMzXvdE5i7zxVMkRJAlsXPn-2ZSvdzldAJLComsobNch6uWuWWo9MLiVLH9VRrFyRwqWWpibC20WyuWN-kCeDX2uRultzpHpGFqQ_XFSj1nuHRFABG9xSELgYy4v1TUVHadutpVn5TX2ur0_FD-aQAovN26PoRk_)

**Review by nicholascar — APPROVED** [review]
Thanks @aucampia, as always! I think the diagram really helps and it looks really great. I think I'll do more of these in other Python projects of mine that have interesting class hierarchies (I am actually a data modeller, not a software engineer after all).

Sadly, looking at the diagram, I think we need to tidy up the classes more:

I think we should update `URIRef` to a more naturally-named `IRI` class - while retaining an alias so people can still call `URIRef` in code, if that's possible - and also consider renaming `Identifier` to `Term` to avoid overuse of the word _identifier_, while retaining a legacy alias too. Like the potential `ConjunctiveGraph` / `Dataset` deduplication, these structures and names matter for they cloud understanding if bad and help it greatly if good.

Something for rdflib 7.0.0...

Waiting for a review from @gjhiggins since he made the first class diagram.

**Review by gjhiggins — APPROVED** [review]
Puts my clumsy effort to shame, well worthwhile addition.

**Comment by gjhiggins** [comment]
Documentation note - `Variable` is noticeably missing from the docs. The docstring is laconic:
```py
class Variable(Identifier):
    """
    A Variable - this is used for querying, or in Formula aware
    graphs, where Variables can stored in the graph
    """
```
But there are examples of the former and the W3's “Notation 3” document has a [section devoted to Formulae](https://www.w3.org/TeamSubmission/n3/#Quoting) which can usefully be mined for content to flesh out a description of the latter.

**Comment by nicholascar** [comment]
I had to make a small update to get the diagram rendering: tab in all in so Sphinx, at least on my computer, could interpret it correctly. Fix in a new PR I'm about to make for the multiple namespaces issue.

**Comment by aucampia** [comment]
We should probably do some manner of publishing of the site from master so we can also spot potential issues with docs before a release.

**Comment by nicholascar** [comment]
We do already have `master` being built over at readthedocs into `latest` documentation, as opposed to the latest release being built into `stabel`, so we should easily be able to see what the next release’ documentation will look like before a release

**Comment by edmondchuc** [comment]
Maybe build a docs preview on each PR. You can see how it works here. https://github.com/ternaustralia/linkeddata-site/pull/85. And here is the GitHub Actions workflow. https://github.com/ternaustralia/linkeddata-site/blob/master/.github/workflows/preview.yml. It utilises the free surge.sh service for serving the static site content. 

Alternatively, [render.com](https://render.com/) also has a generous free tier with PR previews for static sites as well.

**Comment by aucampia** [comment]
> Maybe build a docs preview on each PR. You can see how it works here. [ternaustralia/linkeddata-site#85](https://github.com/ternaustralia/linkeddata-site/pull/85). And here is the GitHub Actions workflow. https://github.com/ternaustralia/linkeddata-site/blob/master/.github/workflows/preview.yml. It utilises the free surge.sh service for serving the static site content.
> 
> Alternatively, [render.com](https://render.com/) also has a generous free tier with PR previews for static sites as well.

Thanks for this, I will have a look, will be very helpful to do.

**Comment by aucampia** [comment]
> We do already have `master` being built over at readthedocs into `latest` documentation, as opposed to the latest release being built into `stable`, so we should easily be able to see what the next release’ documentation will look like before a release

I had a look at https://readthedocs.org/projects/rdflib/builds/ and it seems latests was not built recently, what triggers an update there?

**Comment by nicholascar** [comment]
> what triggers an update there?

At the moment, I think it's all manual, as in someone has to go in an press 'build' for `latest`, `stable` and release builds (e.g. `5.0.0`) when a new release is made.

Auto-building would be better but I don't know how to auto-build in reathedocs but I'm sure it's possible as I think many PyPI projects auto-build.

**Comment by edmondchuc** [comment]
@aucampia I just noticed RTD also provides pull request previews. See https://docs.readthedocs.io/en/stable/pull-requests.html

**Pull request merged**
